bin/consequence_mapping/run_consequence_mapping.sh

```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+# A wrapper script for running the consequence mapping pipeline concurrently and based on a VCF input.
+# See /docs/consequence-mapping.md for detailed information.
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+bcftools query "$1" -f '%CHROM:%POS:%REF:%ALT\n' \
+  | sort -u \
+  | parallel --pipe -j 20 -N 200 python $DIR/consequence_mapping.py \
```

Comment:
Could you please consult Ensembl VEP devs to ensure that they don't have any concurrency rate limiting implemented for their web service (like we do for ours)?

Reply:
I am not well-versed with the failure semantics for parallel... If one of the concurrent python processes fails midway, are those outputs discarded or the pipe completely fails? the python program uses logger which prints out info and error messages... will all of those stdout/stderr outputs (presumably created in parallel) be captured coherently without race conditions? Also, I am curious if you had a chance to test the shell script by invoking it with bsub...

Reply:
To answer your questions:
1. I already consulted [the relevant documentation](https://github.com/Ensembl/ensembl-rest/wiki#example-clients) concerning rate limiting. They allow submitting 15 requests per second. Given that each instance of a `consequence_mapping.py` script submits only a single requests, and it takes on average 10 seconds to process, the effective request rate for 20 workers running in parallel is about 2 requests per second—a great deal less than the allowed limit.
2. Internally, `consequence_mapping.py` script catches request errors using the `@retry` decorator and attempts to retry them up to 10 times, with increasing intervals (plus some jitter). If after all attempts a request was still unsuccessful, the script will fail with non-zero return code.
3. If at least one process handled by GNU Parallel fails, the entire run will fail with non-zero return code. This is expected behaviour, because if at least chunk of data was not processed, the entire run should fail.
4. Concerning stream redirection:
  + STDOUT of `consequence_mapping.py` is redirected by `run_consequence_mapping.sh` into a specific file and is not further processed by Parallel or bsub.
  + STDERR of `consequence_mapping.py` flows through `run_consequence_mapping.sh` unchanged, is caught by Parallel (and further by bsub, if it is used).
5. I have performed end-to-end testing and checked the assertions in points 1-4 work—the Ensembl API errors are being handled and STDOUT/STDERR/script output are being correctly handled by bsub

Reply:
Given that you have performed end-to-end testing already, could you include running the shell script as part of the test suite too? Parallelism is notoriously hard to get right and since it looks like we have gotten it right now, having safeguards at the test suite will guard against future regressions...

Reply:
Yes, to have at least an integration test is a good idea. I will implement this in a separate ticket, since we will be moving this code into a separate repository and this will require a separate task anyway.